### PR TITLE
Add support for the CellCam Kikker camera.

### PR DESCRIPTION
**Pull request by marsiwiec**
Hi,

I added support for the CellCam Kikker camera. The trick was to add a getExposure() call because the MicroManager driver does not expose this as a property. The other changes are just downstream from that. Everything seems to be working on my end. Hope this helps!

Marcin

### CODE REVIEW

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 1]
```diff
@@ -426,7 +439,9 @@ def getParam(self, param):
             'bitDepth': 'PixelType',
         }.get(param, param)
         with self.camLock:
-            val = self.mmc.getProperty(self.camName, str(paramTrans))
+            if paramTrans == 'Exposure':
+                val = self.mmc.getExposure(self.camName) # workaround for CellCam
+                val = self.mmc.getProperty(self.camName, str(paramTrans))
```
Comment (outofculture) [inline acq4/devices/MicroManagerCamera/mmcamera.py:448]: This line probably needs to be in an else block?

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 2]
```diff
@@ -152,21 +157,26 @@ def _readAllParams(self):
...
             for prop in properties:
                 vals = self.mmc.getAllowedPropertyValues(self.camName, prop)
                 if vals == ():
-                    if self.mmc.hasPropertyLimits(self.camName, prop):
-                        vals = (
-                            self.mmc.getPropertyLowerLimit(self.camName, prop),
-                            self.mmc.getPropertyUpperLimit(self.camName, prop),
-                        )
+                    if prop != 'Exposure': # workaround for exposure working
+                        if self.mmc.hasPropertyLimits(self.camName, prop):
```
Comment (outofculture) [inline acq4/devices/MicroManagerCamera/mmcamera.py:165]: This branch of logic is still valid for non-CellCam cameras.

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 3]
```diff
@@ -152,21 +157,26 @@ def _readAllParams(self):
...
+                                self.mmc.getPropertyLowerLimit(self.camName, prop),
+                                self.mmc.getPropertyUpperLimit(self.camName, prop),
+                            )
+                        else:
+                            # just guess..
+                            vals = (1e-6, 1e3)
                     else:
-                        # just guess..
-                        vals = (1e-6, 1e3)
+                        vals = (1, 100) # sensible range of exposure values...
```
Comment (outofculture) [inline acq4/devices/MicroManagerCamera/mmcamera.py:174]: 😢  I wish we didn't have to guess. This is a fine guess, though.

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 4]
```diff
@@ -152,21 +157,26 @@ def _readAllParams(self):
...
                 else:
                     vals = list(vals)
-                readonly = self.mmc.isPropertyReadOnly(self.camName, prop)
+                if prop != 'Exposure': # again, workaround...
+                    readonly = self.mmc.isPropertyReadOnly(self.camName, prop)
```
Comment (outofculture) [inline acq4/devices/MicroManagerCamera/mmcamera.py:178]: This is still valid code for non-CellCam cameras.

Comment (outofculture) [reply acq4/devices/MicroManagerCamera/mmcamera.py:178]: Oh, also, `readonly` is used in code further along, so we would need a default value (I'm guessing `False`) for CellCam.

**Review by outofculture — CHANGES_REQUESTED** [review]
These mostly all seem alright, but I'd like to reduce the chance of this code altering the behavior of other cameras. To that end, I would just add `and self.camName == "CellCam"` to all the if statements regarding `"Exposure"`. Let me know if you want to do that, or I can just take this over, no problem.

File: acq4/devices/MicroManagerCamera/mmcamera.py (hunk 4 above)
Comment (marsiwiec) [inline acq4/devices/MicroManagerCamera/mmcamera.py:178]:
```suggestion
                    readonly = self.mmc.isPropertyReadOnly(self.camName, prop)
                else:
                    readonly = False
```

Comment (marsiwiec) [reply acq4/devices/MicroManagerCamera/mmcamera.py:178]: Yeah, so we could perhaps add an else block.
```suggestion
                    readonly = self.mmc.isPropertyReadOnly(self.camName, prop)
                else:
                    readonly = False
```

File: acq4/devices/MicroManagerCamera/mmcamera.py (hunk 3 above)
Comment (marsiwiec) [reply acq4/devices/MicroManagerCamera/mmcamera.py:174]: Yeah, it sucks with the guesses. If only they would add exposure to properties...

File: acq4/devices/MicroManagerCamera/mmcamera.py (hunk 2 above)
Comment (marsiwiec) [reply acq4/devices/MicroManagerCamera/mmcamera.py:165]: ok!

File: acq4/devices/MicroManagerCamera/mmcamera.py (hunk 1 above)
Comment (marsiwiec) [reply acq4/devices/MicroManagerCamera/mmcamera.py:448]: You're right, my bad!
```suggestion
            else:
                val = self.mmc.getProperty(self.camName, str(paramTrans))
```

**Review by marsiwiec — COMMENTED** [review]
Hi @outofculture, I made the suggested changes, let me know if this is ok.

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 5]
```diff
@@ -391,7 +401,10 @@ def _setParam(self, param, value, autoCorrect=True):
 
         with self.camLock:
             for param, value in setParams:
-                self.mmc.setProperty(self.camName, str(param), str(value))
+                if param == 'Exposure': # workaround for CellCam - call to setExposure(), not getProperty()
```
Comment (marsiwiec) [inline acq4/devices/MicroManagerCamera/mmcamera.py]:
```suggestion
                if param == 'Exposure' and self.camName == "CellCam": # workaround for CellCam - call to setExposure(), not getProperty()
```

File: acq4/devices/MicroManagerCamera/mmcamera.py [hunk 6]
```diff
@@ -426,7 +439,9 @@ def getParam(self, param):
             'bitDepth': 'PixelType',
         }.get(param, param)
         with self.camLock:
-            val = self.mmc.getProperty(self.camName, str(paramTrans))
+            if paramTrans == 'Exposure':
```
Comment (marsiwiec) [inline acq4/devices/MicroManagerCamera/mmcamera.py]:
```suggestion
            if paramTrans == 'Exposure' and self.camName == "CellCam":
```

**Pull request merged**